pkg/storage/unified/resource/server.go

```diff
@@ -0,0 +1,537 @@
+
+	// Real storage backend
+	Search ResourceSearchServer
+
+	// Diagnostics
+	Diagnostics DiagnosticsServer
+
+	// Check if a user has access to write folders
+	// When this is nil, no resources can have folders configured
+	WriteAccess WriteAccessHooks
+
+	// Callbacks for startup and shutdown
+	Lifecycle LifecycleHooks
+}
+
+func NewResourceServer(opts ResourceServerOptions) (ResourceServer, error) {
+	if opts.Tracer == nil {
+		opts.Tracer = noop.NewTracerProvider().Tracer("resource-server")
+	}
+
+	if opts.NextEventID == nil {
+		if opts.NodeID == 0 {
+			opts.NodeID = rand.Int63n(1024)
+		}
+		eventNode, err := snowflake.NewNode(opts.NodeID)
+		if err != nil {
+			return nil, apierrors.NewInternalError(
+				fmt.Errorf("error initializing snowflake id generator :: %w", err))
+		}
+		opts.NextEventID = func() int64 {
+			return eventNode.Generate().Int64()
+		}
+	}
+
+	if opts.Store == nil {
+		return nil, fmt.Errorf("missing AppendingStore implementation")
+	}
+	if opts.Search == nil {
+		opts.Search = &noopService{}
+	}
+	if opts.Diagnostics == nil {
+		opts.Search = &noopService{}
+	}
+
+	return &server{
+		tracer:      opts.Tracer,
+		nextEventID: opts.NextEventID,
+		store:       opts.Store,
+		search:      opts.Search,
+		diagnostics: opts.Diagnostics,
+		access:      opts.WriteAccess,
+		lifecycle:   opts.Lifecycle,
+	}, nil
+}
+
+var _ ResourceServer = &server{}
+
+type server struct {
+	tracer      trace.Tracer
+	nextEventID func() int64
+	store       AppendingStore
+	search      ResourceSearchServer
+	diagnostics DiagnosticsServer
+	access      WriteAccessHooks
+	lifecycle   LifecycleHooks
+
+	// init checking
+	once    sync.Once
+	initErr error
+}
+
+// Init implements ResourceServer.
+func (s *server) Init() error {
+	s.once.Do(func() {
+		// TODO, setup a broadcaster for watch
+
+		// Call lifecycle hooks
+		if s.lifecycle != nil {
+			err := s.lifecycle.Init()
+			if err != nil {
+				s.initErr = fmt.Errorf("initialize Resource Server: %w", err)
+			}
+		}
+	})
+	return s.initErr
+}
+
+func (s *server) Stop() {
+	s.initErr = fmt.Errorf("service is stopping")
+	if s.lifecycle != nil {
+		s.lifecycle.Stop()
+	}
+	s.initErr = fmt.Errorf("service is stopped")
+}
+
+func (s *server) newEvent(ctx context.Context, key *ResourceKey, value, oldValue []byte) (*WriteEvent, error) {
+	if key.Name == "" {
+		return nil, apierrors.NewBadRequest("empty name")
+	}
+	// TODO -- make sure it is alphanumeric+
+
+	var err error
+	event := &WriteEvent{
+		EventID: s.nextEventID(),
+		Key:     key,
+		Value:   value,
+	}
+	event.Requester, err = identity.GetRequester(ctx)
+	if err != nil {
+		return nil, ErrUserNotFoundInContext
+	}
+
+	dummy := &metav1.PartialObjectMetadata{}
+	err = json.Unmarshal(value, dummy)
+	if err != nil {
+		return nil, ErrUnableToReadResourceJSON
+	}
+
+	obj, err := utils.MetaAccessor(dummy)
+	if err != nil {
+		return nil, apierrors.NewBadRequest("invalid object in json")
+	}
+	if obj.GetUID() == "" {
```

Comment:
Trusting the caller that the GUID makes me a bit nervous.

spitballing.

Given the server is generic, we could allow it to manage some of the metadata fields. Things like UID, potentially add annotations like CreatedAt, UpdatedAt, ResourceVersion, or other attributes common across resources (one that I have in mind is an has_blob annotation, or etag/size). It feels that there are good reasons for the storage to "own" some object meta fields.

However, the AppendingStore should NOT allow any update to the payload.